src/main/java/org/kiwiproject/dropwizard/util/health/MongoHealthCheck.java

```diff
@@ -1,60 +1,119 @@
+            var isHealthy = ok(okValue);
+            var resultBuilder = newResultBuilder(isHealthy).withDetail("ok", okValue);
 
             if (isHealthy) {
-                return resultBuilder.withMessage("Mongo %s/%s is up", host, dbName).build();
+                return resultBuilder
+                    .withMessage("Successfully pinged Mongo database %s", dbName)
+                    .build();
             }
 
+            var errorMessage = errorMessage(commandResult).orElse(NO_ERRMSG_VALUE);
             return resultBuilder
-                    .withMessage("Failed to retrieve db stats %s/%s : %s",
-                            host, dbName, statsResult.getErrorMessage())
-                    .build();
+                .withMessage("Error pinging Mongo database %s : %s", dbName, errorMessage)
+                .withDetail("code", commandResult.get("code"))
+                .withDetail("codeName", commandResult.get("codeName"))
+                .build();
+
         } catch (Exception e) {
-            LOG.error("Unable to connect to Mongo: {}/{}", host, dbName, e);
-            return newUnhealthyResult(HealthStatus.CRITICAL, "Unable to connect to Mongo %s/%s: %s",
-                    host, dbName, e.getMessage());
+            var message = f("Error pinging Mongo database {} ({}: {})",
+                    dbName, e.getClass().getName(), e.getMessage());
+            LOG.error(message, e);
+            return newUnhealthyResult(e, message);
+        }
+    }
+
+    /**
+     * @implNote The old Mongo 3.x driver had a CommandResult class that encapsulated the
+     * logic to check if commands were successful. It checked whether the 'ok' property in the
+     * result Document was a boolean, which is why this method is checking for both numeric
+     * and boolean values. The Mongo docs say that commands always return a 0 or 1 (integer)
+     * as the value of the 'ok' property, but in reality it returns 0.0 (double), so I decided
+     * to leave the original logic to be conservative. However, since the result should be a
+     * number, this performs the check for a number first, whereas the CommandResult checked
+     * for a boolean first.
+     */
+    private static boolean ok(Object okValue) {
+        if (okValue instanceof Number) {
+            return ((Number) okValue).intValue() == 1;
+        } else if (okValue instanceof Boolean) {
+            return (boolean) okValue;
+        } else {
+            LOG.warn("Received unexpected {} value in ping result for 'ok': {}",
+                    classNameOrNull(okValue),okValue);
+            return false;
         }
     }
+
+    private static String classNameOrNull(Object value) {
+        return Optional.ofNullable(value)
+                .map(Object::getClass)
+                .map(Class::getName)
+                .orElse(null);
+    }
+
+    private static Optional<String> errorMessage(Document commandResult) {
+        var errmsgValue = commandResult.get("errmsg");
```

Comment:
Does `Map` have a method that returns `Optional`? Or, could just use the `getOrDefault` method? In that case, rename this to something like `errorMessageOrDefault`

Reply:
It's fine the way it is...